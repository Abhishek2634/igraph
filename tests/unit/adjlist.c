/*
   IGraph library.
   Copyright (C) 2021  The igraph development team <igraph@igraph.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include <igraph.h>

#include "test_utilities.inc"

int test_simple_trees() {
    igraph_t g, g2;
    igraph_adjlist_t adjlist;
    igraph_bool_t iso;

    /* Directed, out */
    igraph_tree(&g, 42, 3, IGRAPH_TREE_OUT);
    igraph_adjlist_init(&g, &adjlist, IGRAPH_OUT, IGRAPH_LOOPS_ONCE, IGRAPH_MULTIPLE);
    igraph_adjlist(&g2, &adjlist, IGRAPH_OUT, /*duplicate=*/ 0);
    igraph_isomorphic(&g, &g2, &iso);
    if (!iso) {
        return 1;
    }
    igraph_adjlist_destroy(&adjlist);
    igraph_destroy(&g2);
    igraph_destroy(&g);

    /* Directed, in */
    igraph_tree(&g, 42, 3, IGRAPH_TREE_OUT);
    igraph_adjlist_init(&g, &adjlist, IGRAPH_IN, IGRAPH_LOOPS_ONCE, IGRAPH_MULTIPLE);
    igraph_adjlist(&g2, &adjlist, IGRAPH_IN, /*duplicate=*/ 0);
    igraph_isomorphic(&g, &g2, &iso);
    if (!iso) {
        return 1;
    }
    igraph_adjlist_destroy(&adjlist);
    igraph_destroy(&g2);
    igraph_destroy(&g);

    /* Undirected */
    igraph_tree(&g, 42, 3, IGRAPH_TREE_UNDIRECTED);
    igraph_adjlist_init(&g, &adjlist, IGRAPH_OUT, IGRAPH_LOOPS_ONCE, IGRAPH_MULTIPLE);
    igraph_adjlist(&g2, &adjlist, IGRAPH_ALL, /*duplicate=*/ 1);
    igraph_isomorphic(&g, &g2, &iso);
    if (!iso) {
        return 1;
    }
    igraph_adjlist_destroy(&adjlist);
    igraph_destroy(&g2);
    igraph_destroy(&g);

    return 0;
}

int main() {
    int retval;

    RUN_TEST(test_simple_trees);

    return 0;
}
