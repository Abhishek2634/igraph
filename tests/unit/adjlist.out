Testing loop edge elimination in undirected graph

Loops eliminated: {
  0: ( 1 3 3 )
  1: ( 0 2 )
  2: ( 1 3 )
  3: ( 0 0 2 4 )
  4: ( 3 )
}

Loops listed once: {
  0: ( 1 3 3 )
  1: ( 0 2 )
  2: ( 1 2 3 )
  3: ( 0 0 2 4 )
  4: ( 3 4 4 )
}

Loops listed twice: {
  0: ( 1 3 3 )
  1: ( 0 2 )
  2: ( 1 2 2 3 )
  3: ( 0 0 2 4 )
  4: ( 3 4 4 4 4 )
}

============================================================

Testing loop edge elimination in directed graph

In-edges, loops eliminated: {
  0: ( 3 )
  1: ( 0 )
  2: ( 1 )
  3: ( 0 2 )
  4: ( 3 )
}

In-edges, loops listed once: {
  0: ( 3 )
  1: ( 0 )
  2: ( 1 2 )
  3: ( 0 2 )
  4: ( 3 4 4 )
}

In-edges, loops listed once even if IGRAPH_LOOPS_TWICE is given: {
  0: ( 3 )
  1: ( 0 )
  2: ( 1 2 )
  3: ( 0 2 )
  4: ( 3 4 4 )
}

Out-edges, loops eliminated: {
  0: ( 1 3 )
  1: ( 2 )
  2: ( 3 )
  3: ( 0 4 )
  4: ( )
}

Out-edges, loops listed once: {
  0: ( 1 3 )
  1: ( 2 )
  2: ( 2 3 )
  3: ( 0 4 )
  4: ( 4 4 )
}

Out-edges, loops listed once even if IGRAPH_LOOPS_TWICE is given: {
  0: ( 1 3 )
  1: ( 2 )
  2: ( 2 3 )
  3: ( 0 4 )
  4: ( 4 4 )
}

In- and out-edges, loops eliminated: {
  0: ( 1 3 3 )
  1: ( 0 2 )
  2: ( 1 3 )
  3: ( 0 0 2 4 )
  4: ( 3 )
}

In- and out-edges, loops listed once: {
  0: ( 1 3 3 )
  1: ( 0 2 )
  2: ( 1 2 3 )
  3: ( 0 0 2 4 )
  4: ( 3 4 4 )
}

In- and out-edges, loops listed twice: {
  0: ( 1 3 3 )
  1: ( 0 2 )
  2: ( 1 2 2 3 )
  3: ( 0 0 2 4 )
  4: ( 3 4 4 4 4 )
}

============================================================

